### PR TITLE
IDE-660 ECL IDE should allow for an expired password to be changed

### PR DESCRIPTION
**Pull request by dehilsterlexis**
IDE-660 ECL IDE should allow for an expired password to be changed

### CODE REVIEW

**Comment by dehilsterlexis** [comment]
@GordonSmith for review. It turns out that an expired password does not return when MyAccount is called. It works for non-expired passwords only. So we still have not solved the problem for expired passwords.

**Comment by richardkchapman** [comment]
This should be targeting 7.0.8 I would think

**Comment by dehilsterlexis** [comment]
@GordonSmith please take a look at this. I have put in buttons for change password. I think this design pretty good but I'd like your opinion before going through a thorough test if you think this has gone too far.

File: comms/Account.h [hunk 1]
```diff
@@ -3,4 +3,4 @@
 #include "comms.h"
 
 COMMS_API bool ChangePassword(const CString &serverIP, const CString &user, const CString &oldPassword, const CString &newPassword, const CString &confirmPassword, int &retCode, CString &retMsg);
-COMMS_API bool VerifyUser(IConfig * config, const CString &user, const CString &password, int &retCode, CString &retMsg);
+COMMS_API bool VerifyUser(IConfig * config, const CString &user, const CString &password, int &retCode, CString &retMsg, int &passwordExpirationWarningDays, int &passwordDaysRemaining);
```
Comment (GordonSmith) [inline comms/Account.h]: minor: prefix new params with "ret" for consistency

File: en_us/en_us.rc [hunk 2]
```diff
@@ -507,22 +507,23 @@ BEGIN
...
-    DEFPUSHBUTTON   "OK",IDOK,101,78,50,14
-    PUSHBUTTON      "Cancel",IDCANCEL,153,78,50,14
-    CTEXT           "",IDC_STATIC_UPGRADE,7,58,195,16
+    EDITTEXT        IDC_EDIT_PASSWORD,61,40,162,14,ES_PASSWORD | ES_AUTOHSCROLL
+    CTEXT           "",IDC_STATIC_MSG,64,58,154,16
+    PUSHBUTTON      "Pre&ferences",IDC_LOGIN_PREFS,4,78,55,14,WS_DISABLED
+    DEFPUSHBUTTON   "OK",IDOK,117,78,50,14
+    PUSHBUTTON      "Cancel",IDCANCEL,172,78,50,14
+    CTEXT           "",IDC_STATIC_UPGRADE,61,58,141,16
+    PUSHBUTTON      "Change Psswd",IDC_BUTTON_CHANGEPASSWORD,4,61,54,14
```
Comment (GordonSmith) [inline en_us/en_us.rc]: Screenshot?

File: en_us/en_us.rc [hunk 3]
```diff
@@ -560,6 +561,8 @@ BEGIN
     PUSHBUTTON      "Pre&ferences...",IDC_LOGIN_PREFS_RHS,231,120,54,14
     DEFPUSHBUTTON   "OK",IDOK,331,144,50,14
     PUSHBUTTON      "Cancel",IDCANCEL,383,144,50,14
+    PUSHBUTTON      "Change Password",IDC_CHANGEPASSWORD,73,120,63,14
+    PUSHBUTTON      "Change Password",IDC_CHANGEPASSWORD_RHS,290,120,63,14
```
Comment (GordonSmith) [inline en_us/en_us.rc]: Screenshot?

File: wlib/LoginDlg.cpp [hunk 4]
```diff
@@ -179,17 +181,34 @@ class CLoginDlg : public CDialogImpl<CLoginDlg>,
...
 			//prime the repository cache with the new credentials
 			CString server(m_config->Get(GLOBAL_SERVER_ATTRIBUTE));
 			CComPtr<IRepository> rep = ::AttachRepository(server, m_User, m_Password, m_ConfigLabel);
 			EndDialog(nID);
+
+			if (m_User.GetLength() && passwordDaysRemaining) {
+				if (passwordDaysRemaining == 1)
+					retMsg.Format(_T("Your password expires in %i day"), passwordDaysRemaining, retMsg);
+				else
+					retMsg.Format(_T("Your password expires in %i days"), passwordDaysRemaining, retMsg);
```
Comment (GordonSmith) [inline wlib/LoginDlg.cpp]: IMO overkill - I would just use:  `Your password expires in %i day(s)`

File: wlib/LoginDlg.cpp [hunk 5]
```diff
@@ -699,6 +730,24 @@ class CLoginDlgAMT : public CDialogImpl<CLoginDlgAMT>,
 		return 0;
 	}
 
+	LRESULT OnBnClickedChangePassword(WORD /*wNotifyCode*/, WORD /*wID*/, HWND /*hWndCtl*/, BOOL& /*bHandled*/)
+	{
+		DoDataExchange(true);
+		GetIConfig(QUERYBUILDER_CFG)->Set(GLOBAL_USER, m_UserLHS);
+		GetIConfig(QUERYBUILDER_CFG)->Set(GLOBAL_SERVER_ACCOUNT, m_configLHS->Get(GLOBAL_SERVER_ATTRIBUTE));
+		DoModalChangePassword();
```
Comment (GordonSmith) [inline wlib/LoginDlg.cpp]: Has this been tested?

File: wlib/LoginDlg.cpp [hunk 6]
```diff
@@ -699,6 +730,24 @@ class CLoginDlgAMT : public CDialogImpl<CLoginDlgAMT>,
...
+		return 0;
+	}
+
+	LRESULT OnBnClickedChangePasswordRHS(WORD /*wNotifyCode*/, WORD /*wID*/, HWND /*hWndCtl*/, BOOL& /*bHandled*/)
+	{
+		DoDataExchange(true);
+		GetIConfig(QUERYBUILDER_CFG)->Set(GLOBAL_USER, m_UserRHS);
+		GetIConfig(QUERYBUILDER_CFG)->Set(GLOBAL_SERVER_ACCOUNT, m_configRHS->Get(GLOBAL_SERVER_ATTRIBUTE));
+		DoModalChangePassword();
```
Comment (GordonSmith) [inline wlib/LoginDlg.cpp]: Has this been tested?

File: comms/Account.h (hunk 1 above)
Comment (dehilsterlexis) [reply comms/Account.h]: Done

File: en_us/en_us.rc (hunk 2 above)
Comment (dehilsterlexis) [reply en_us/en_us.rc]: amtlogin01

File: en_us/en_us.rc (hunk 3 above)
Comment (dehilsterlexis) [reply en_us/en_us.rc]: amtlogin01

File: en_us/en_us.rc (hunk 2 above)
Comment (dehilsterlexis) [reply en_us/en_us.rc]: login01

File: wlib/LoginDlg.cpp (hunk 6 above)
Comment (dehilsterlexis) [reply wlib/LoginDlg.cpp]: Half way. Yes, I need to test this. I was first wanting a general feel from you before finishing.

File: en_us/en_us.rc (hunk 2 above)
Comment (dehilsterlexis) [reply en_us/en_us.rc]: @GordonSmith made this longer. Will combine the other jira as a subtask which people were complaining about the small size of the configuration pulldown.

**Comment by GordonSmith** [comment]
@dehilsterlexis lets discuss this a bit more, but I am thinking (having slept on it):
* Move the change password button to the "server" tab in preferences
* After login and when there are less than N days left until password expires, popup a dialog asking them if they would like to change the PW now (deliberately annoying).

**Comment by dehilsterlexis** [comment]
@GordonSmith ready for review. I would like if you could actually run this yourself for a second set of eyes.

**Comment by dehilsterlexis** [comment]
@GordonSmith I found one problem. I will let you know when it is fixed.

File: comms/Account.cpp [hunk 7]
```diff
@@ -85,15 +87,39 @@ COMMS_API bool VerifyUser(IConfig * config, const CString &user, const CString &
...
+            _ns1__MyAccountRequest request;
+            _ns1__MyAccountResponse response;
+            ESP_EXCEPTION_LOG3(response.Exceptions);
+            if (server.MyAccount(&request, &response) == SOAP_OK)
+            {
+                SAFE_ASSIGN(retPasswordExpirationWarningDays, response.passwordExpirationWarningDays);
+                SAFE_ASSIGN(retPasswordDaysRemaining, response.passwordDaysRemaining);
+
+                CString responseMessage;
+                if (responseMessage.IsEmpty() && response.Exceptions)
```
Comment (GordonSmith) [inline comms/Account.cpp]: responseMessage will always be empty here?

File: comms/Account.cpp [hunk 8]
```diff
@@ -85,15 +87,39 @@ COMMS_API bool VerifyUser(IConfig * config, const CString &user, const CString &
...
+                        if (response.Exceptions->Exception[i]->Message)
+                        {
+                            if (!retMsg.IsEmpty())
+                                retMsg += _T("\r\n");
+                            retMsg += response.Exceptions->Exception[i]->Message->c_str();
+                        }
+                    }
+                }
+                return true;
+            }
```
Comment (GordonSmith) [inline comms/Account.cpp]: Move this code into its own method and call it from here.

**Comment by dehilsterlexis** [comment]
@GordonSmith for review. I did leave in two small fixes in AMT that were not graying out the username / password textboxes correctly.  I just thought: leave it in. But I can take them out if you like.

File: comms/Account.cpp (hunk 8 above)
Comment (GordonSmith) [reply comms/Account.cpp]: @dehilsterlexis bump

Comment (dehilsterlexis) [reply comms/Account.cpp]: @GordonSmith done

Comment (dehilsterlexis) [reply comms/Account.cpp]: @GordonSmith will do

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: comms/Account.cpp [hunk 9]
```diff
@@ -8,6 +8,78 @@
...
+
+        return NoErrorResponse(response, retMsg);
+    }
+    else
+    {
+        PasswordError(server, config, retCode, retMsg);
+    }
+    return false;
+}
+
```
Comment (GordonSmith) [inline comms/Account.cpp]: What you had previously was cleaner than this?  All I was suggesting was moving the MyAccount call into its own function? (Note:  it doesn't need all the extra VerifyUser initialization and tear down - its just a regular SOAP call like all the other ones):
```c++
COMMS_API bool MyAccount(IConfig * config, const CString &user, const CString &password, int &retPasswordExpirationWarningDays, int &retPasswordDaysRemaining)
{
	CSoapInitialize<ws_USCOREaccountServiceSoapProxy> server(config->Get(GLOBAL_SERVER_ACCOUNT), user, password);
	_ns1__MyAccountRequest request;
	_ns1__MyAccountResponse response;
	ESP_EXCEPTION_LOG3(response.Exceptions);
	if (server.MyAccount(&request, &response) == SOAP_OK)
	{
		SAFE_ASSIGN(retPasswordExpirationWarningDays, response.passwordExpirationWarningDays);
		SAFE_ASSIGN(retPasswordDaysRemaining, response.passwordDaysRemaining);
		return true;
	}
	else
		_DBGLOG(config->Get(GLOBAL_SERVER_ACCOUNT), LEVEL_WARNING, server.GetClientErrorMsg());

	return false;
}
```

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

**Comment by dehilsterlexis** [comment]
@GordonSmith for review. It turns out that the responses are all different classes:

_ns1__VerifyUserResponse
_ns1__MyAccountResponse
_ns1__UpdateUserResponse

I suggest we leave it that way for now unless you have a better suggestion.

**Comment by GordonSmith** [comment]
Yes but `response.Exceptions` all have the same data type (I think)?

File: comms/Account.cpp [hunk 10]
```diff
@@ -67,33 +118,18 @@ COMMS_API bool VerifyUser(IConfig * config, const CString &user, const CString &
     ESP_EXCEPTION_LOG3(response.Exceptions);
     if (server.VerifyUser(&request, &response) == SOAP_OK)
     {
-        SAFE_ASSIGN(retCode, response.retcode);
-        CString responseMessage;
-#if _COMMS_VER < 300800
```
Comment (dehilsterlexis) [inline comms/Account.cpp]: @GordonSmith I removed the ifdef because _ns1__UpdateUserResponse is the only response that has std::wstring *message;

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: comms/Account.cpp [hunk 11]
```diff
@@ -8,6 +8,56 @@
...
+
+COMMS_API bool MyAcountPasswordInfo(IConfig * config, const CString &user, const CString &password, int &retCode, CString &retMsg, int &retPasswordExpirationWarningDays, int &retPasswordDaysRemaining)
+{
+	CSoapInitialize<ws_USCOREaccountServiceSoapProxy> server(config->Get(GLOBAL_SERVER_ACCOUNT), user, password);
+	if (server.namespaces == NULL)	//  No Server  ---
+	{
+		retCode = 1003;
+		retMsg = _T("Unable to communicate with server.");
+		return false;
+	}
```
Comment (GordonSmith) [inline comms/Account.cpp]: This code is not needed - its only needed in VerifyUser.

File: comms/Account.cpp [hunk 12]
```diff
@@ -8,6 +8,56 @@
...
+            {
+                if (!retMsg.IsEmpty())
+                    retMsg += _T("\r\n");
+
+                retMsg += exceptions->Exception[i]->Message->c_str();
+                noExceptions = false;
+            }
+        }
+    }
+    return noExceptions;
```
Comment (GordonSmith) [inline comms/Account.cpp]: `GetExceptionResponse` should return true IF it got an exception response and false otherwise.

Comment (dehilsterlexis) [reply comms/Account.cpp]: @GordonSmith agreed

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: comms/Account.cpp [hunk 13]
```diff
@@ -8,6 +8,56 @@
...
+
+COMMS_API bool MyAcountPasswordInfo(IConfig * config, const CString &user, const CString &password, int &retCode, CString &retMsg, int &retPasswordExpirationWarningDays, int &retPasswordDaysRemaining)
+{
+    CSoapInitialize<ws_USCOREaccountServiceSoapProxy> server(config->Get(GLOBAL_SERVER_ACCOUNT), user, password);
+    if (server.namespaces == NULL)	//  No Server  ---
+    {
+        retCode = 1003;
+        retMsg = _T("Unable to communicate with server.");
+        return false;
+    }
```
Comment (GordonSmith) [inline comms/Account.cpp]: This code is not needed - its only needed in VerifyUser.

**Comment by dehilsterlexis** [comment]
@GordonSmith for review. Tested including changing the password and it is all working.

**Comment by GordonSmith** [comment]
@dehilsterlexis see above comment.

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

**Comment by GordonSmith** [comment]
Code looks good.

**Comment by GordonSmith** [comment]
@richardkchapman this can be pulled.

**Pull request merged**